gpu/plchain.cu

```diff
@@ -483,84 +552,328 @@ void plchain_debug_analysis(stream_ptr_t stream){
+        kernel_mem_usage[uid] = (float)batch_n/stream_setup.max_anchors_stream*100;
+#endif // DEBUG_PRINT
+        // // sanity check
+        // if (stream_setup.max_anchors_stream < total_n){
+        //     fprintf(stderr, "max_anchors_stream %lu total_n %lu n_read %d\n",
+        //             stream_setup.max_anchors_stream, total_n, n_read);
+        // }
+
+        // if (stream_setup.max_range_grid < griddim) {
+        //     fprintf(stderr, "max_range_grid %d griddim %d, total_n %lu n_read %d\n",
+        //             stream_setup.max_range_grid, griddim, total_n, n_read);
+        // }
+
+        assert(stream_setup.max_anchors_stream >= batch_n);
+        assert(stream_setup.max_range_grid >= griddim);
+        assert(stream_setup.max_num_cut >= cut_num);
+        // work on micro batch
+#ifdef USEHIP
+        roctxRangePushA("reorg");
+#endif
+        // step1: reorg input
+        plmem_reorg_input_arr(reads + read_start, read_end - read_start,
+                          &stream_setup.streams[stream_id].host_mems[uid],
+                          range_kernel_config);
+        // step2: copy to device
+#ifdef USEHIP
+        roctxRangePop();
+#endif
+
+#ifdef DEBUG_PRINT
+        cudaEventRecord(stream_setup.streams[stream_id].short_kernel_start_event[uid],
+                    stream_setup.streams[stream_id].cudastream);
+#endif // DEBUG_PRINT
+        plmem_async_h2d_short_memcpy(&stream_setup.streams[stream_id], uid);
+        // step3: range selection
+        plrange_async_range_selection(&stream_setup.streams[stream_id].dev_mem,
+                                    &stream_setup.streams[stream_id].cudastream);
+        // step4: score generation for short and mid segs
+        plscore_async_short_mid_forward_dp(&stream_setup.streams[stream_id].dev_mem,
+                                    &stream_setup.streams[stream_id].cudastream);
+        // step5: copy short and mid results back
+        plmem_async_d2h_short_memcpy(&stream_setup.streams[stream_id], uid);
+#ifdef DEBUG_PRINT
+        cudaEventRecord(stream_setup.streams[stream_id].short_kernel_stop_event[uid],
+                    stream_setup.streams[stream_id].cudastream);
+#endif // DEBUG_PRINT
+        // update index
+        read_start = read_end;
+
+#ifdef USEHIP
+        roctxRangePop();
+#endif
+
+#ifdef DEBUG_CHECK
+        plchain_debug_analysis_short(stream_setup.streams[stream_id], uid, kernel_throughput);
+#endif // DEBUG_CHECK
     }
 
-    if (stream_setup.max_range_grid < griddim) {
-        fprintf(stderr, "max_range_grid %d griddim %d, total_n %lu n_read %d\n",
-                stream_setup.max_range_grid, griddim, total_n, n_read);
+// FIXME: temporary solution for microbatching
+    if (read_start < n_read) {
+        fprintf(stderr, "[WARNING] Unable to fit reads %d - %d into a microbatch. Fall back to cpu chaining\n", read_start, n_read-1);
     }
```

Comment:
Temporary fix: somethings the total number of anchors < MICRO_BATCH * batch_size, but cannot be divided into microbatches without cutting a read into two halves. If 1 or 2 reads are left behind because of this, we use CPU kenel in map.c. In the future, different threads creates different microbatch, and is responsible for making sure this does not happen.